tendermint/src/rpc/event_listener.rs

```diff
@@ -57,102 +57,85 @@ impl EventListener {
-                    Ok(json_value) => Ok(Event::GenericJSONEvent(json_value)),
-                    Err(_) => Ok(Event::GenericStringEvent(msg.to_string())),
-                },
-            },
-        }
+        let result_event =
+            serde_json::from_str::<WrappedResultEvent>(&msg.to_string())?.into_result()?;
+
+        Ok(result_event.data)
     }
 }
-/// The Event enum is typed events emmitted by the Websockets
+
+// TODO: (later) this should live somewhere else; these events are also
+// published byte the event bus independent from RPC.
+// We leave it here for now because unsupported types are still
+// decodeable via fallthrough variants (GenericJSONEvent).
+/// The Event enum is typed events emitted by the Websockets
 #[derive(Serialize, Deserialize, Debug, Clone)]
-pub enum Event {
-    /// The result of the ABCI app processing a transaction, serialized as JSON RPC response
-    JsonRPCBlockResult(
-        /// The Block Result
-        Box<RPCBlockResult>,
-    ),
+#[serde(tag = "type", content = "value")]
+#[allow(clippy::large_enum_variant)]
+pub enum TMEventData {
+    /// EventDataNewBlock is returned upon subscribing to "tm.event='NewBlock'"
+    #[serde(alias = "tendermint/event/NewBlock")]
+    EventDataNewBlock(EventDataNewBlock),
 
-    /// The result of the ABCI app processing a transaction, serialized as JSON RPC response
-    JsonRPCTransactionResult(
-        /// the tx result data
-        Box<RPCTxResult>,
-    ),
+    /// EventDataTx is returned upon subscribing to "tm.event='Tx'"
+    #[serde(alias = "tendermint/event/Tx")]
+    EventDataTx(EventDataTx),
 
     ///Generic event containing json data
     GenericJSONEvent(
         /// generic event json data
         serde_json::Value,
     ),
-    ///Generic String Event
-    GenericStringEvent(
-        /// generic string data
-        String,
-    ),
 }
 
-/// Websocket result for Processed Transactions
-pub type JsonRPCTransactionResult = Wrapper<RPCTxResult>;
-
-/// Websocket result for Processed Block
-pub type JsonRPCBlockResult = Wrapper<RPCBlockResult>;
-
-/// JSON RPC Result Type
+/// Event data from a subscription
 #[derive(Serialize, Deserialize, Debug, Clone)]
-pub struct RPCTxResult {
-    query: String,
-    data: Data,
-    events: HashMap<String, Vec<String>>,
+pub struct ResultEvent {
+    query: Option<String>,
+    data: Option<TMEventData>,
+    events: Option<HashMap<String, Vec<String>>>,
 }
-impl response::Response for RPCTxResult {}
+impl response::Response for ResultEvent {}
 
-/// TX data
-#[derive(Serialize, Deserialize, Debug, Clone)]
-struct Data {
-    #[serde(rename = "type")]
-    data_type: String,
-    value: TxValue,
-}
+/// JSONRPC wrapped ResultEvent
+pub type WrappedResultEvent = Wrapper<ResultEvent>;
 
 /// TX value
 #[derive(Serialize, Deserialize, Debug, Clone)]
-struct TxValue {
+pub struct EventDataTx {
     #[serde(rename = "TxResult")]
```

Comment:
This could be a the container attribute `rename_all = "PascalCase"`.

Reply:
I'd rather keep those renamings local to the var. Seems more consistent to the remaining code.